pkg/deps/maven.go

```diff
@@ -0,0 +1,591 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+package deps
+
+import (
+	"archive/zip"
+	"bufio"
+	"bytes"
+	"encoding/xml"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"golang.org/x/net/html/charset"
+
+	"github.com/apache/skywalking-eyes/license-eye/internal/logger"
+	"github.com/apache/skywalking-eyes/license-eye/pkg/license"
+)
+
+type MavenPomResolver struct {
+	maven string
+	repo  string
+}
+
+// CanResolve determine whether the file can be resolve by name of the file
+func (resolver *MavenPomResolver) CanResolve(mavenPomFile string) bool {
+	base := filepath.Base(mavenPomFile)
+	logger.Log.Debugln("Base name:", base)
+	return base == "pom.xml"
+}
+
+// Resolve resolves licenses of all dependencies declared in the pom.xml file.
+func (resolver *MavenPomResolver) Resolve(mavenPomFile string, report *Report) error {
+	if err := os.Chdir(filepath.Dir(mavenPomFile)); err != nil {
+		return err
+	}
+
+	if err := resolver.CheckMVN(); err != nil {
+		return err
+	}
+
+	deps, err := resolver.LoadDependencies()
+	if err != nil {
+		return err
+	}
+
+	if err := resolver.ResolveDependencies(deps, report); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// CheckMVN check available maven tools, find local repositories and download all dependencies
+func (resolver *MavenPomResolver) CheckMVN() error {
+	if err := resolver.FindMaven("./mvnw"); err == nil {
+		logger.Log.Debugln("mvnw is found, will use mvnw by default")
+	} else if err := resolver.FindMaven("mvn"); err != nil {
+		return fmt.Errorf("neither found mvnw nor mvn")
+	}
+
+	logger.Log.Debugln("mvnw is not found, will use mvn by default")
+
+	if err := resolver.FindLocalRepository(); err != nil {
+		return fmt.Errorf("can not find the local repository: %v", err)
+	}
+
+	if err := resolver.DownloadDeps(); err != nil {
+		return fmt.Errorf("dependencies download error")
+	}
+
+	return nil
+}
+
+func (resolver *MavenPomResolver) FindMaven(execName string) error {
+	if _, err := exec.Command(execName, "--version").Output(); err != nil {
+		return err
+	}
+
+	resolver.maven = execName
+	return nil
+}
+
+func (resolver *MavenPomResolver) FindLocalRepository() error {
+	output, err := exec.Command(resolver.maven, "help:evaluate", "-Dexpression=settings.localRepository", "-q", "-DforceStdout").Output() // #nosec G204
+	if err != nil {
+		return err
+	}
+
+	resolver.repo = string(output)
+	return nil
+}
+
+func (resolver *MavenPomResolver) DownloadDeps() error {
+	cmd := exec.Command(resolver.maven, "process-resources") // #nosec G204
```

Comment:
This seems to be problematic in multi-modules maven project, can you try this feature in our main repo http://github.com/apache/skywalking ?

Reply:
as you said in discussion, the failure may be caused by submodules. so i resolve it using maven command `install -DskipTests` to download possible dependencies, and it may take some time.